app/models/reduction_fetcher.rb

```diff
@@ -35,9 +35,23 @@ def reduce_by_subject?
     @topic == :reduce_by_subject
   end
 
-  def where(query)
-    return @subject_reductions.where(query.except(:user_id)) if reduce_by_subject?
-    return @user_reductions.where(query.except(:subject_id)) if reduce_by_user?
+  def where(selector, in_place: false)
+    if in_place
+      return locate_in_place(selector.except(:user_id), @subject_reductions, SubjectReduction) if reduce_by_subject?
+      return locate_in_place(selector.except(:subject_id), @user_reductions, UserReduction) if reduce_by_user?
+    else
+      return @subject_reductions.where(selector.except(:user_id)).first_or_initialize if reduce_by_subject?
+      return @user_reductions.where(selector.except(:subject_id)).first_or_initialize if reduce_by_user?
+    end
+  end
+
+  def locate_in_place(selector, relation, factory)
+    match = relation.to_a.map(&:serializable_hash).select{ |record| key_match(record, selector) }
+    match.empty? ? factory.new(selector) : match[0]
+  end
+
+  def key_match(record, selector)
+    selector.all? { |key, _ | record[key] == selector[key]}
```

Comment:
this confused me and I may not quite understand what it's trying to do but I think this can be changed to `selector.all? { |key, value| record[key] == value }`? Or am i way off?

Reply:
done